stdlib: add List.split_map

craigfe opened the pull request:
Adds the following functions to the standard library:

```ocaml
module List : sig
  (* ... *)

  val split_with   : ('a -> 'b * 'c) -> 'a t -> 'b t * 'c t
  val combine_with : ('a -> 'b -> 'c) -> 'a t -> 'b t -> 'c t
end
```

These functions generalise the existing `List.split` and `List.combine` to take arbitrary functions for splitting and combining. In particular:

```ocaml
List.split    ≡  List.split_with (fun (a, b) -> (a, b))
List.combine  ≡  List.combine_with (fun a b -> (a, b))
```

These functions are useful when dealing with data in records or abstract types rather than pairs, and I'm occasionally bitten by the lack of them when refactoring the latter into the former. Providing them here avoids the need for a two-pass transformation w/ `List.map` in these cases.

These functions are quite popular in Haskell code – as [`zipWith`](https://hoogle.haskell.org/?hoogle=zipWith) and [`unzipWith`](https://hoogle.haskell.org/?hoogle=unzipWith) – but AFAICS are not provided by the big OCaml stdlib replacement libraries; make of that what you will.

craigfe committed:
stdlib: add List.split_map

gasche commented:
`List.combine_with` is already available as `List.map2`; I am not convinced that we need another name for the function.

`List.split_with` sounds useful, but I'm not sure about the name:
- `_with` is not used for this purpose in the stdlib so far (when it does occur at the end of a name, it denotes an argument: `compare_length_with`, `ends_with`, `starts_with`)
- the Haskell folks use `By` (`_by`), so they may find `_by` more familiar (why did you choose `_with`?)

(This is not a criticism about the name per se, but rather a request for rationale on the naming so that we can compare with other proposals, for example `split_by` and `split_map`.)

Note: Ideally we should have not just one name, but a family of name that works at higher arities too (`'a -> 'b1 * 'b2 * 'b3`, etc.); I think that `split3_with` works okay for this.

nojb commented:
> `List.split_with` sounds useful, but I'm not sure about the name:

What about `List.split_map` ? (EDIT: OK, I see you suggested this name already, apologies)

craigfe commented:
> `List.combine_with` is already available as `List.map2`; I am not convinced that we need another name for the function.

Thanks, I somehow missed `map2`... Agreed, I'll drop it from the PR.

> the Haskell folks use `By` (`_by`), so they may find `_by` more familiar (why did you choose `_with`?)

The Haskell equivalents of these functions are `zipWith` and `unzipWith`. There are other examples of this pattern in Haskell's core libraries, including [`insertWith`](https://hoogle.haskell.org/?hoogle=insertWith), [`unionWith`](https://hoogle.haskell.org/?hoogle=unionWith), [`intersectionWith`](https://hoogle.haskell.org/?hoogle=intersectionWith), etc. I think the [Haskell idiom of `fooBy`](https://hackage.haskell.org/package/base-4.14.1.0/docs/Data-List.html#g:23) captures a different idea of replacing a typeclass constraint with an explicit argument, as in:

```haskell
sort   :: Ord a => [a] -> [a]
sortBy :: (a -> a -> Ordering) -> [a] -> [a]

union   :: Eq a => [a] -> [a] -> [a]
unionBy :: (a -> a -> Bool) -> [a] -> [a] -> [a] 
```

We can also consider a distinction between `With` and `By` in that the former generally takes a data transformation to be lifted and the latter takes a projection onto a type with a desired property. Personally, I find that `split_with f` reads more intuitively than `split_by f`, making it clearer that `f` is the function actually doing the splitting rather than providing a _property_ necessary for splitting (c.f. ``sortBy (compare `on` fst)`` in Haskell).

I'm not especially against `split_map`, but find it less intuitive and worry that "map" is already quite overloaded w.r.t. lists.

> Note: Ideally we should have not just one name, but a family of name that works at higher arities too (`'a -> 'b1 * 'b2 * 'b3`, etc.); I think that `split3_with` works okay for this.

Agreed. FWIW, Haskell's stdlib [also contains](https://hackage.haskell.org/package/base-4.14.1.0/docs/Data-List.html#v:zipWith3) `zipWith3`, `zipWith4`, ..., `zipWith7` :-)

craigfe commented:
Dropped `List.combine_with` and added a small doc-comment to `List.combine` indicating that it's generalised by `List.map2`,  people like me who missed it :slightly_smiling_face:

damiendoligez commented:
The `split_map` name is very natural because
```
let split_map f l = split (map f l)
```
You can see the same idea in `concat_map` and (somewhat less clearly) in `filter_map` and `fold_left_map`.

craigfe commented:
@damiendoligez: Good point, I've switched to that :+1:

(FTR, I still find the convention of `List.foo_map` to be overloaded – e.g. `fold_left_map` isn't a very natural composition of `fold_left` and `map` – but it does fit in this case.)

xvw commented on stdlib/list.mli:
```diff
@@ -429,10 +429,25 @@ val split : ('a * 'b) list -> 'a list * 'b list
    [combine [a1; ...; an] [b1; ...; bn]] is
    [[(a1,b1); ...; (an,bn)]].
+
+   Equivalent to [map2 ~f:(fun a b -> (a, b))].
```
I think that here the label should not be present.

craigfe replied:
This file is auto-generated from `listLabels.mli`, so a compromise is needed somewhere; I'll switch it so that `ListLabels` is the incorrect one.

xvw replied:
Thanks a lot!

nojb commented:
This function was added in #14185. Closing!
